Add lint to check manual pattern char comparison

AurelienFT opened the pull request:
This PR adds a new lint asked in https://github.com/rust-lang/rust-clippy/issues/12490

This lint catches manual char comparison in pattern of string functions and propose to use `char` or array of `char` instead.

As it's my first contribution i'm not feeling very safe about not matching too much or missing some cases.

Thanks in advance for taking time to review and propose feedback !

changelog: new lint [`manual_pattern_char_comparison`]

Alexendoo commented on clippy_lints/src/methods/manual_pattern_char_comparison.rs:
```diff
@@ -0,0 +1,148 @@
+            && args.len() > pos
+            && let arg = args[pos]
+            && let ExprKind::Closure(closure) = arg.kind
+            && closure.fn_decl.inputs.len() == 1
+            && let TyKind::Path(QPath::Resolved(_, c_arg_type)) = closure.fn_decl.inputs[0].kind
+            && c_arg_type.res == Res::PrimTy(PrimTy::Char)
```
This part prevents e.g. `.find(|c| ...)` without an explicit type `c: char` being linted, a direct way to check the type would be  `cx.typeck_results().pat_ty(body.params[0].pat)`

However since a closure `Pattern` always takes a `char` so we can just remove this check 

```suggestion
```

AurelienFT replied:
The problem that I have is that my lint isn't restricted to methods on string and so if I remove these lines it trigger [here](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/search_is_some_fixable_none.rs#L57) which is a closure with an `usize` (following what I understand from the code and the debug I made). Is there a way to make sure my lint is triggered on string/str in a proper way or is it better to use the `cx.typeck_results().pat_ty(body.params[0].pat)` you proposed in case there is closure with a character in others function ?

Alexendoo replied:
`single_char_patterns` checks that the receiver is a string so you'd inherit that when merging them

https://github.com/rust-lang/rust-clippy/blob/5aae5f6ae6b4ba0e6ddadea43e6a2bcadcaa2328/clippy_lints/src/methods/single_char_pattern.rs#L45-L46

That's better than checking with `pat_ty` because we don't want to suggest to replace say `"abc".chars().find(|c| c == 'a')` with `"abc".chars.find('a')`. It would be good to add that as a test case

Alexendoo replied:
`Iterator::find` gives you a reference to the item so that example doesn't work, but a test such as

```rust
struct NotStr;

impl NotStr {
    fn find(&self, _: impl FnMut(char) -> bool) {}
}
```

would still be good

AurelienFT replied:
I have added the test !

AurelienFT commented on clippy_lints/src/methods/manual_pattern_char_comparison.rs:
```diff
@@ -0,0 +1,148 @@
+
+use super::MANUAL_PATTERN_CHAR_COMPARISON;
+
+const PATTERN_METHODS: [(&str, usize); 22] = [
+    ("contains", 0),
+    ("starts_with", 0),
+    ("ends_with", 0),
+    ("find", 0),
+    ("rfind", 0),
+    ("split", 0),
+    ("split_inclusive", 0),
+    ("rsplit", 0),
+    ("split_terminator", 0),
+    ("rsplit_terminator", 0),
+    ("splitn", 1),
+    ("rsplitn", 1),
+    ("split_once", 0),
+    ("rsplit_once", 0),
+    ("matches", 0),
+    ("rmatches", 0),
+    ("match_indices", 0),
+    ("rmatch_indices", 0),
+    ("trim_start_matches", 0),
+    ("trim_end_matches", 0),
+    ("replace", 0),
+    ("replacen", 0),
+];
```
I'm not sure to understand what you mean by : `move both of these lints to their own lint pass`. Do you mean make them use the same `check` function ? If that's the case how should I organize ? Call this check from the other with the list as parameter ? Move code of both in the same file ?

Alexendoo replied:
A new file for both outside of `methods/`, so a new file at e.g. `clippy_lints/src/string_patterns.rs` and move both the lints into it

```rust
declare_clippy_lint! {
    ...
    pub MANUAL_PATTERN_CHAR_COMPARISON,
    ...
}

declare_clippy_lint! {
    ...
    pub SINGLE_CHAR_PATTERN,
    ...
}

declare_lint_pass!(StringPatterns => [MANUAL_PATTERN_CHAR_COMPARISON, SINGLE_CHAR_PATTERN]);

impl<'tcx> LateLintPass<'tcx> for StringPatterns {
    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
        if !expr.span.from_expansion()
            && let ExprKind::MethodCall(..) = expr.kind
            // the common bits
        {

        }
    }
}
```

You may need to then `cargo dev update_lints` and register `StringPatterns` in `clippy_lints/src/lib.rs`

Alexendoo replied:
Here `StringPatterns` would be the lint pass containing the two lints

AurelienFT replied:
Ok thank for the clear explanation and the walkthrough. I'm in the process of merging them where you proposed but I have an issue : the lint `single_char_patterns` use a function called `get_hint_if_single_char_arg` which is also use by others lint in the `methods/` folder but we can't have access in the parent folder.

I didn't this an `..._utils.rs` file in `clippy_lints/src` where I was confident enough to place this function. Do you have an idea of the emplacement I could place this function in ?

AurelienFT replied:
I merged both as you asked and placed the util function in the `clippy_utils` crate.

AurelienFT commented on tests/ui/search_is_some.rs:
```diff
@@ -73,6 +74,7 @@ fn is_none() {
     // check that we don't lint if `find()` is called with
     // `Pattern` that is not a string
+    #[allow(clippy::manual_pattern_char_comparison)]
     let _ = "hello world".find(|c: char| c == 'o' || c == 'l').is_none();
```
Done :)

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,223 @@
+                        ControlFlow::Break(())
+                    }
+                },
+                ExprKind::Match(match_value, arms, _) => {
+                    if snippet(cx, match_value.span, "") != param_name.name.as_str() || arms.len() != 2 {
+                        return ControlFlow::Break(());
+                    }
```
A nit, destructure `arms` here as `[arm, _]` rather than checking the len + indexing

We should check that `sub_expr` is an expansion of `matches!()` so that we don't let through other things. We can do that with `matching_root_macro_call(cx, sub_expr.span, sym::matches_macro).is_some()`. For example

```rust
"".find(|c| match c {
    'a' | 'b' => true,
    _ => c.is_ascii(),
});
```

We also need to check that `arm` has no guard:

```rust
"".find(|c| matches!(c, 'a' | 'b' if false));
```

AurelienFT replied:
Applied. I learnt a lot of helper ! :o

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,223 @@
+                    if snippet(cx, match_value.span, "") != param_name.name.as_str() || arms.len() != 2 {
+                        return ControlFlow::Break(());
+                    }
+                    if let PatKind::Or(pats) = arms[0].pat.kind {
+                        for pat in pats {
+                            if let PatKind::Lit(expr) = pat.kind
+                                && let ExprKind::Lit(lit) = expr.kind
+                                && let LitKind::Char(c) = lit.node
+                            {
+                                set_chars.push(format!("'{}'", c.escape_default()));
+                            }
+                        }
+                    } else if let PatKind::Lit(expr) = arms[0].pat.kind
+                        && let ExprKind::Lit(lit) = expr.kind
+                        && let LitKind::Char(c) = lit.node
+                    {
+                        set_chars.push(format!("'{}'", c.escape_default()));
+                    }
+                    ControlFlow::Continue(Descend::No)
```
If we see a pattern that isn't a literal we need to avoid linting also, e.g.

```rust
"".find(|c| matches!(c, 'a' | '1'..'4'));
"".find(|c| c == 'a' || matches!(c, '1'..'4'));
```

`walk_short` can make it a little nicer by combining the `Or`/`Lit` cases, if it returns `false` we'd need to return `Break`

```rust
arm.pat.walk_short(|pat| match pat.kind {
    PatKind::Lit(expr) if let ExprKind::Lit.. => { /* push */ }
    PatKind::Or(_) => true,
    _ => false,
});
```

AurelienFT replied:
`walk_short` makes it so clearer. Thank you !

Alexendoo commented on clippy_utils/src/str_utils.rs:
```diff
@@ -289,6 +296,49 @@ pub fn to_camel_case(item_name: &str) -> String {
     s
 }
 
+/// Return an hint if the `arg` is a string with a single character
+pub fn get_hint_if_single_char_arg(
+    cx: &LateContext<'_>,
+    arg: &Expr<'_>,
+    applicability: &mut Applicability,
+    ascii_only: bool,
+) -> Option<String> {
```
Let's rename this and expand the description a bit

```suggestion
/// Converts `expr` to a `char` literal if it's a `str` literal containing a single character (or a single byte with `ascii_only`)
pub fn str_literal_to_char_literal(
    cx: &LateContext<'_>,
    expr: &Expr<'_>,
    applicability: &mut Applicability,
    ascii_only: bool,
) -> Option<String> {
```

Alexendoo replied:
I think it belongs more in `clippy_utils::source` also, this module is for operations on the non syntax kind of `str`s

AurelienFT replied:
Ok thanks for the explanations. Applied

AurelienFT commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,223 @@
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
```
I applied your logic just complexified it a bit because `c` could be LHS or RHS. Thanks for the clear explanations

AurelienFT commented:
I have applied all the suggestions. Thank you for the very deep explanation as a first contributor it's very appreciable. I learnt a lot of things in Clippy and it motivate me to continue to contribute !

AurelienFT commented:
@Alexendoo thank you for the heart, that's kind ! If there anything I need to do for this PR or I need to wait ? I don't want to be rude, it's because I'm not very familiar with the process so I prefer to ask

Alexendoo commented:
I plan to do another read through of it soon, then after that I'll create a final comment period thread on Zulip for others to weigh in

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,239 @@
+            match sub_expr.kind {
+                ExprKind::Binary(op, left, right) if op.node == BinOpKind::Eq => {
+                    if path_to_local_id(left, binding)
+                        && cx.typeck_results().expr_ty_adjusted(right).kind() == &ty::Char
+                        && let Some(c) = get_char_value(right)
```
can use `.is_char()` over `kind() == Char`, I think it makes sense to move the type check into `get_char_value` also

AurelienFT replied:
Done

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+}
+
+fn get_char_value(expr: &Expr<'_>) -> Option<String> {
+    match expr.kind {
+        ExprKind::Lit(lit) => match lit.node {
+            LitKind::Char(c) => Some(format!("'{}'", c.escape_default())),
+            _ => None,
+        },
+        ExprKind::Path(QPath::Resolved(_, path)) => {
+            if path.segments.len() == 1 {
+                let segment = &path.segments[0];
+                if segment.args.is_none() {
+                    Some(segment.ident.name.to_string())
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
```
A check here for `expr.from_expansion()` would be good so we don't suggest crossing any macro boundaries e.g.

```rust
macro_rules! m {
    ($e:expr) => {
        $e == '?'
    };
}
"".find(|c| m!(c));
```

For determining which exprs we can suggest for you could use [`switch_to_eager_eval`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/eager_or_lazy/fn.switch_to_eager_eval.html)

AurelienFT replied:
Added a test case for this. However `switch_to_eager_eval` doesn't seems to work : 

- the macro call return `true`,
- `(|c: char| c == '.' || c == ',' || c == '!')` return `true`

We are not able to make the distinction, so I used `expr.span.from_expansion()`

Alexendoo replied:
You'd need both as they do separate things, the `.from_expansion` rules out macro expanded expressions

The `switch_to_eager_eval` is like an extended version of the current `match`, e.g. it will include `foo.bar` as long as there's no user defined `Deref` going on

AurelienFT replied:
Nice ! ty done.

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,239 @@
+                },
+                ExprKind::Binary(op, _, _) if op.node == BinOpKind::Or => ControlFlow::Continue(Descend::Yes),
+                ExprKind::Match(match_value, [arm, _], _) => {
+                    if matching_root_macro_call(cx, sub_expr.span, sym::matches_macro).is_none() {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.guard.is_some() {
+                        return ControlFlow::Break(());
+                    }
+                    if !path_to_local_id(match_value, binding) {
+                        return ControlFlow::Break(());
+                    }
+                    if arm.pat.walk_short(|pat| match pat.kind {
```
These `if`s can be merged, or if it makes the indentation weird at least the first 3 can

AurelienFT replied:
Done on the three first. The fourth one is more complex and I find that merge them isn't more readable.

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,239 @@
+            && let Some(&(_, pos)) = PATTERN_METHODS
+                .iter()
+                .find(|(array_method_name, _)| *array_method_name == method_name)
+            && args.len() > pos
+        {
+            let arg = &args[pos];
```
`.get()` here rather than index + len check

Alexendoo commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,239 @@
+        {
+            return;
+        }
+        match set_chars.len().cmp(&1) {
+            Ordering::Equal => span_lint_and_sugg(
+                cx,
+                MANUAL_PATTERN_CHAR_COMPARISON,
+                method_arg.span,
+                "this manual char comparison can be written more succinctly",
+                "consider using a `char`",
+                set_chars[0].clone(),
+                applicability,
+            ),
+            Ordering::Greater => span_lint_and_sugg(
+                cx,
+                MANUAL_PATTERN_CHAR_COMPARISON,
+                method_arg.span,
+                "this manual char comparison can be written more succinctly",
+                "consider using an array of `char`",
+                format!("[{}]", set_chars.join(", ")),
+                applicability,
+            ),
+            Ordering::Less => {},
+        }
```
Could use a single `span_lint_and_then` here, would also let us skip some allocation happening when the lint is allowed (the callback is not called if that is the case)

Alexendoo commented:
Opened the FCP here also: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/FCP.20.60manual_pattern_char_comparison.60

bors commented:
:umbrella: The latest upstream changes (presumably #12815) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Alexendoo commented on clippy_lints/src/string_patterns.rs:
````diff
@@ -0,0 +1,244 @@
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for manual char comparison in string patterns
+    ///
+    /// ### Why is this bad?
+    /// This can be written more concisely using a `char` or an array of `char`.
+    /// This is more readable and more optimized when comparing to only one `char`.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// "Hello World !".trim_end_matches(|c: char| c == '.' || c == ',' || c == '!' || c == '?');
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// "Hello World !".trim_end_matches(['.', ',', '!', '?']);
````
Some doc nits

```suggestion
    /// Checks for manual `char` comparison in string patterns.
    ///
    /// ### Why is this bad?
    /// This can be written more concisely using a `char` or an array of `char`s.
    /// This is more readable and more optimized when comparing to only one `char`.
    ///
    /// ### Example
    /// ```no_run
    /// "Hello World!".trim_end_matches(|c| c == '.' || c == ',' || c == '!' || c == '?');
    /// ```
    /// Use instead:
    /// ```no_run
    /// "Hello World!".trim_end_matches(['.', ',', '!', '?']);
```

AurelienFT replied:
Done

AurelienFT commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,244 @@
+}
+
+fn check_manual_pattern_char_comparison(cx: &LateContext<'_>, method_arg: &Expr<'_>) {
+    let applicability = Applicability::MachineApplicable;
```
Done

AurelienFT commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,244 @@
+    let applicability = Applicability::MachineApplicable;
+    if let ExprKind::Closure(closure) = method_arg.kind
+        && let body = cx.tcx.hir().body(closure.body)
+        && let PatKind::Binding(_, binding, ..) = body.params[0].pat.kind
```
Done

AurelienFT commented on clippy_lints/src/string_patterns.rs:
```diff
@@ -0,0 +1,244 @@
+                    diag.span_suggestion(
+                        method_arg.span,
+                        "consider using a `char`",
+                        snippet_opt(cx, set_char_spans[0]).unwrap(),
```
Done

AurelienFT commented:
I addressed all the comments.

Furthermore I saw that last time I forget to push modification on function `get_char_span` that reduce the code to less than 10 lines thanks to your ideas @Alexendoo.

Everything should be fine now

bors commented:
:umbrella: The latest upstream changes (presumably #12915) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Alexendoo commented:
If you could squash the commits I think this is ready to merge, the FCP has been running long enough without issue

AurelienFT commented:
@Alexendoo Seems ok to me. (I'm not very confident with my git skills 😆 )

Alexendoo commented:
Yep, looks great thanks!

@bors r+

bors commented:
:pushpin: Commit c86b19f1ef262eeb1992a91879b38ffbed07a1a7 has been approved by `Alexendoo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=Alexendoo c86b19f1ef262eeb1992a91879b38ffbed07a1a7 -->
<!-- homu: {"type":"Approved","sha":"c86b19f1ef262eeb1992a91879b38ffbed07a1a7","approver":"Alexendoo","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit c86b19f1ef262eeb1992a91879b38ffbed07a1a7 with merge 6f1b05bad08b31d4f5de05bbb0c7ba917d98e6bd...
<!-- homu: {"type":"BuildStarted","head_sha":"c86b19f1ef262eeb1992a91879b38ffbed07a1a7","merge_sha":"6f1b05bad08b31d4f5de05bbb0c7ba917d98e6bd"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/9473297649/job/26100635298)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/actions/runs/9473297649/job/26100635298","builder_name":"checks-action_test"} -->

Alexendoo commented:
@bors retry

bors commented:
:hourglass: Testing commit c86b19f1ef262eeb1992a91879b38ffbed07a1a7 with merge 740b72ef79f1453c1ef76af3b58e13e10fcb32de...
<!-- homu: {"type":"BuildStarted","head_sha":"c86b19f1ef262eeb1992a91879b38ffbed07a1a7","merge_sha":"740b72ef79f1453c1ef76af3b58e13e10fcb32de"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/9473478572/job/26101238650), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/9473478587/job/26101251125), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/9473478570/job/26101463758)
Approved by: Alexendoo
Pushing 740b72ef79f1453c1ef76af3b58e13e10fcb32de to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Alexendoo","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9473478570/job/26101463758","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9473478572/job/26101238650","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/9473478587/job/26101251125"},"merge_sha":"740b72ef79f1453c1ef76af3b58e13e10fcb32de"} -->

AurelienFT commented:
@Alexendoo I want to make you a special thanks for being so nice an pedagogy for this PR, loved it. You are really kind !

Alexendoo commented:
Thank you! Congrats on the first contribution
